File: python_project/algorithms/bresenham.py
"""
Bresenham's Line Drawing Algorithm.
Uses integer-only decision parameters and supports all 8 octants.
"""

class BresenhamLine:
    def __init__(self, x1: int, y1: int, x2: int, y2: int):
        self.x1 = int(x1)
        self.y1 = int(y1)
        self.x2 = int(x2)
        self.y2 = int(y2)

    def generate_steps(self):
        dx = abs(self.x2 - self.x1)
        dy = abs(self.y2 - self.y1)
        sx = 1 if self.x1 < self.x2 else -1
        sy = 1 if self.y1 < self.y2 else -1

        is_steep = dy > dx
        num_steps = dy if is_steep else dx

        if num_steps == 0:
            return [{
                "step": 0,
                "total_steps": 0,
                "pixel": (self.x1, self.y1),
                "pk": 0,
                "candidate1": (self.x1, self.y1),
                "candidate2": (self.x1, self.y1),
                "selected": (self.x1, self.y1),
                "decision_cond": "Single Point",
                "explanation": f"Start and end points coincide at ({self.x1}, {self.y1}).",
                "pixels": [(self.x1, self.y1)]
            }]

        cur_x = self.x1
        cur_y = self.y1
        history = [(cur_x, cur_y)]
        step_records = []

        if not is_steep:
            # Shallow slope |m| <= 1: X is driving axis
            pk = 2 * dy - dx
            two_dy = 2 * dy
            two_dy_minus_two_dx = 2 * dy - 2 * dx

            step_records.append({
                "step": 0,
                "total_steps": num_steps,
                "pixel": (cur_x, cur_y),
                "pk": pk,
                "candidate1": (cur_x + sx, cur_y),
                "candidate2": (cur_x + sx, cur_y + sy),
                "selected": (cur_x, cur_y),
                "decision_cond": f"Initial p0 = 2*dy - dx = 2({dy}) - {dx} = {pk}",
                "explanation": f"Initial pixel plotted at ({cur_x}, {cur_y}). Driving axis is X. p0 = {pk}.",
                "pixels": list(history)
            })

            for i in range(1, num_steps + 1):
                cand_straight = (cur_x + sx, cur_y)
                cand_diagonal = (cur_x + sx, cur_y + sy)

                if pk < 0:
                    chosen = cand_straight
                    next_pk = pk + two_dy
                    cond_str = f"p_{i-1} = {pk} < 0 -> Select ({chosen[0]}, {chosen[1]})"
                    expl = f"pk is negative (< 0) -> Select horizontal neighbor. pk+1 = pk + 2dy = {pk} + {two_dy} = {next_pk}."
                else:
                    chosen = cand_diagonal
                    next_pk = pk + two_dy_minus_two_dx
                    cond_str = f"p_{i-1} = {pk} >= 0 -> Select ({chosen[0]}, {chosen[1]})"
                    expl = f"pk is >= 0 -> Select diagonal neighbor. pk+1 = pk + 2dy - 2dx = {pk} + ({two_dy_minus_two_dx}) = {next_pk}."

                cur_x, cur_y = chosen
                history.append((cur_x, cur_y))

                step_records.append({
                    "step": i,
                    "total_steps": num_steps,
                    "pixel": (cur_x, cur_y),
                    "pk": pk,
                    "pk_next": next_pk,
                    "candidate1": cand_straight,
                    "candidate2": cand_diagonal,
                    "selected": chosen,
                    "decision_cond": cond_str,
                    "explanation": expl,
                    "pixels": list(history)
                })
                pk = next_pk
        else:
            # Steep slope |m| > 1: Y is driving axis
            pk = 2 * dx - dy
            two_dx = 2 * dx
            two_dx_minus_two_dy = 2 * dx - 2 * dy

            step_records.append({
                "step": 0,
                "total_steps": num_steps,
                "pixel": (cur_x, cur_y),
                "pk": pk,
                "candidate1": (cur_x, cur_y + sy),
                "candidate2": (cur_x + sx, cur_y + sy),
                "selected": (cur_x, cur_y),
                "decision_cond": f"Initial p0 = 2*dx - dy = 2({dx}) - {dy} = {pk}",
                "explanation": f"Initial pixel plotted at ({cur_x}, {cur_y}). Driving axis is Y (steep). p0 = {pk}.",
                "pixels": list(history)
            })

            for i in range(1, num_steps + 1):
                cand_straight = (cur_x, cur_y + sy)
                cand_diagonal = (cur_x + sx, cur_y + sy)

                if pk < 0:
                    chosen = cand_straight
                    next_pk = pk + two_dx
                    cond_str = f"p_{i-1} = {pk} < 0 -> Select ({chosen[0]}, {chosen[1]})"
                    expl = f"pk is negative (< 0) -> Select vertical neighbor. pk+1 = pk + 2dx = {pk} + {two_dx} = {next_pk}."
                else:
                    chosen = cand_diagonal
                    next_pk = pk + two_dx_minus_two_dy
                    cond_str = f"p_{i-1} = {pk} >= 0 -> Select ({chosen[0]}, {chosen[1]})"
                    expl = f"pk is >= 0 -> Select diagonal neighbor. pk+1 = pk + 2dx - 2dy = {pk} + ({two_dx_minus_two_dy}) = {next_pk}."

                cur_x, cur_y = chosen
                history.append((cur_x, cur_y))

                step_records.append({
                    "step": i,
                    "total_steps": num_steps,
                    "pixel": (cur_x, cur_y),
                    "pk": pk,
                    "pk_next": next_pk,
                    "candidate1": cand_straight,
                    "candidate2": cand_diagonal,
                    "selected": chosen,
                    "decision_cond": cond_str,
                    "explanation": expl,
                    "pixels": list(history)
                })
                pk = next_pk

        return step_records

File: src/algorithms/bresenham.ts
import { BresenhamStepState, PixelCoord } from '../types';

export function generateBresenhamSteps(x1: number, y1: number, x2: number, y2: number): BresenhamStepState[] {
  const steps: BresenhamStepState[] = [];

  const rawDx = x2 - x1;
  const rawDy = y2 - y1;
  const absDx = Math.abs(rawDx);
  const absDy = Math.abs(rawDy);
  const stepX = rawDx >= 0 ? 1 : -1;
  const stepY = rawDy >= 0 ? 1 : -1;

  const isSteep = absDy > absDx;
  const numSteps = isSteep ? absDy : absDx;

  if (numSteps === 0) {
    const px: PixelCoord = { x: x1, y: y1 };
    return [{
      step: 0,
      totalSteps: 0,
      x: x1,
      y: y1,
      pk: 0,
      dx: 0,
      dy: 0,
      twoDy: 0,
      twoDyMinusTwoDx: 0,
      decisionCondition: 'Single Point',
      chosenCandidate: 'Straight',
      candidate1: px,
      candidate2: px,
      explanation: `Start and end points coincide at (${x1}, ${y1}).`,
      historyPixels: [px]
    }];
  }

  let curX = x1;
  let curY = y1;
  const history: PixelCoord[] = [{ x: curX, y: curY }];

  if (!isSteep) {
    // Shallow line: |m| <= 1, driving axis is X
    const twoDy = 2 * absDy;
    const twoDyMinusTwoDx = 2 * absDy - 2 * absDx;
    let pk = 2 * absDy - absDx;

    // Step 0: Initial point
    steps.push({
      step: 0,
      totalSteps: numSteps,
      x: curX,
      y: curY,
      pk,
      dx: absDx,
      dy: absDy,
      twoDy,
      twoDyMinusTwoDx,
      decisionCondition: `Initial parameter p₀ = 2Δy - Δx = 2(${absDy}) - ${absDx} = ${pk}`,
      chosenCandidate: 'Straight',
      candidate1: { x: curX + stepX, y: curY, isCandidate: true, label: 'E' },
      candidate2: { x: curX + stepX, y: curY + stepY, isCandidate: true, label: 'NE' },
      explanation: `Plotting initial pixel (${curX}, ${curY}). Slope is shallow (|m| ≤ 1). Initial decision parameter p₀ = 2Δy - Δx = ${pk}.`,
      historyPixels: [...history]
    });

    for (let i = 1; i <= numSteps; i++) {
      const candEast: PixelCoord = { x: curX + stepX, y: curY, isCandidate: true, label: stepY >= 0 ? 'East (E)' : 'East-South (ES)' };
      const candNorthEast: PixelCoord = { x: curX + stepX, y: curY + stepY, isCandidate: true, label: stepY >= 0 ? 'North-East (NE)' : 'South-East (SE)' };

      let nextX = curX + stepX;
      let nextY = curY;
      let chosenName: 'East' | 'North-East' = 'East';
      let nextPk = pk;
      let condStr = '';
      let expl = '';

      if (pk < 0) {
        nextY = curY; // East
        nextPk = pk + twoDy;
        chosenName = 'East';
        condStr = `p_${i - 1} = ${pk} < 0 ➔ Select ${candEast.label} (${nextX}, ${nextY})`;
        expl = `Because decision parameter p = ${pk} is negative (< 0), the ideal line is closer to the horizontal neighbor (${nextX}, ${nextY}). Next p = p + 2Δy = ${pk} + ${twoDy} = ${nextPk}.`;
      } else {
        nextY = curY + stepY; // North-East
        nextPk = pk + twoDyMinusTwoDx;
        chosenName = 'North-East';
        condStr = `p_${i - 1} = ${pk} ≥ 0 ➔ Select ${candNorthEast.label} (${nextX}, ${nextY})`;
        expl = `Because decision parameter p = ${pk} is non-negative (≥ 0), the ideal line is closer to diagonal neighbor (${nextX}, ${nextY}). Next p = p + 2Δy - 2Δx = ${pk} + (${twoDyMinusTwoDx}) = ${nextPk}.`;
      }

      curX = nextX;
      curY = nextY;
      history.push({ x: curX, y: curY });

      steps.push({
        step: i,
        totalSteps: numSteps,
        x: curX,
        y: curY,
        pk,
        pkNext: nextPk,
        dx: absDx,
        dy: absDy,
        twoDy,
        twoDyMinusTwoDx,
        decisionCondition: condStr,
        chosenCandidate: chosenName,
        candidate1: candEast,
        candidate2: candNorthEast,
        explanation: expl,
        historyPixels: [...history]
      });

      pk = nextPk;
    }
  } else {
    // Steep line: |m| > 1, driving axis is Y
    const twoDx = 2 * absDx;
    const twoDxMinusTwoDy = 2 * absDx - 2 * absDy;
    let pk = 2 * absDx - absDy;

    // Step 0: Initial point
    steps.push({
      step: 0,
      totalSteps: numSteps,
      x: curX,
      y: curY,
      pk,
      dx: absDx,
      dy: absDy,
      twoDy: twoDx, // using primary axis difference
      twoDyMinusTwoDx: twoDxMinusTwoDy,
      decisionCondition: `Initial parameter p₀ = 2Δx - Δy = 2(${absDx}) - ${absDy} = ${pk}`,
      chosenCandidate: 'Straight',
      candidate1: { x: curX, y: curY + stepY, isCandidate: true, label: 'N' },
      candidate2: { x: curX + stepX, y: curY + stepY, isCandidate: true, label: 'NE' },
      explanation: `Plotting initial pixel (${curX}, ${curY}). Slope is steep (|m| > 1), so Y is the driving axis. Initial p₀ = 2Δx - Δy = ${pk}.`,
      historyPixels: [...history]
    });

    for (let i = 1; i <= numSteps; i++) {
      const candNorth: PixelCoord = { x: curX, y: curY + stepY, isCandidate: true, label: 'North (N)' };
      const candDiagonal: PixelCoord = { x: curX + stepX, y: curY + stepY, isCandidate: true, label: 'Diagonal (NE/NW)' };

      let nextX = curX;
      let nextY = curY + stepY;
      let chosenName: 'Straight' | 'Diagonal' = 'Straight';
      let nextPk = pk;
      let condStr = '';
      let expl = '';

      if (pk < 0) {
        nextX = curX; // North
        nextPk = pk + twoDx;
        chosenName = 'Straight';
        condStr = `p_${i - 1} = ${pk} < 0 ➔ Select Vertical (${nextX}, ${nextY})`;
        expl = `Because p = ${pk} < 0, line is closer to vertical neighbor (${nextX}, ${nextY}). Next p = p + 2Δx = ${pk} + ${twoDx} = ${nextPk}.`;
      } else {
        nextX = curX + stepX; // Diagonal
        nextPk = pk + twoDxMinusTwoDy;
        chosenName = 'Diagonal';
        condStr = `p_${i - 1} = ${pk} ≥ 0 ➔ Select Diagonal (${nextX}, ${nextY})`;
        expl = `Because p = ${pk} ≥ 0, line is closer to diagonal neighbor (${nextX}, ${nextY}). Next p = p + 2Δx - 2Δy = ${pk} + (${twoDxMinusTwoDy}) = ${nextPk}.`;
      }

      curX = nextX;
      curY = nextY;
      history.push({ x: curX, y: curY });

      steps.push({
        step: i,
        totalSteps: numSteps,
        x: curX,
        y: curY,
        pk,
        pkNext: nextPk,
        dx: absDx,
        dy: absDy,
        twoDy: twoDx,
        twoDyMinusTwoDx: twoDxMinusTwoDy,
        decisionCondition: condStr,
        chosenCandidate: chosenName,
        candidate1: candNorth,
        candidate2: candDiagonal,
        explanation: expl,
        historyPixels: [...history]
      });

      pk = nextPk;
    }
  }

  return steps;
}
